### ISSUE TITLE
Multi-project configurations cannot be linted all at once

### ISSUE THREAD

**Issue by benasher44** [dimaMachina/graphql-eslint#2260]
### Issue workflow progress

_Progress of the issue based on the
[Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [x] 1. The issue provides a reproduction available on GitHub, Stackblitz or CodeSandbox

     > For example, you can start off by editng the
     > ['basic' example on Stackblitz](https://stackblitz.com/github/B2o5T/graphql-eslint/tree/master/examples/basic).

     > Please make sure the graphql-eslint version under `package.json` matches yours.

- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

The lint plugin does load per-project schemas, but it loads it once and ends up reusing it for all files. This results in using one project's schema to lint schemas of other projects.

**To Reproduce** Steps to reproduce the behavior:

1. Configure multi-project graphql files setup (e.g. src/project1/**/*.graphql, src/project2/**/*.graphql)
2. Run lint for the whole folder

**Expected behavior**

Lint should pass, but it instead claims that (for example) types are unreachable, since it's checking those types against the project that was loaded for the lint pass.

**Environment:**

- OS: macOS
- `@graphql-eslint/eslint-plugin`: 3.20.1
- Node.js:

**Additional context**

Having looked at the code a little bit, it looks like each rule looks at the loaded schema, which will be the schema loaded by the plugin at the start of the lint pass. Instead, each rule should reuse the code that loads the schema from the project-specific schema cache.

**Comment by benasher44** [comment]
Maybe this is a FlatConfig-specific thing? I can't reproduce this in the example project, but I can definitely reproduce it in ours.

**Comment by benasher44** [comment]
Confirmed this reproduces in the multiple-project sample project. It doesn't reproduce today because both schemas define a User type. If you just change one of them to be called AdminUser instead, it reproduces:

![image](https://github.com/dimaMachina/graphql-eslint/assets/1120429/71f7c83b-7e22-4600-a8ee-f6666ecdc37e)

**Comment by user1736** [comment]
Also run into a similar issue. Any news on the fix?

**Comment by dotansimha** [comment]
@user1736 can you please share the patch fix you made locally? Or what's the suggested workaround for that? @dimaMachina can help with adjusting it.

**Comment by dimaMachina** [comment]
@user1736 hi, could you create a minimal reproduction with steps to reproduce issues with caching?

**Comment by benasher44** [comment]
It's reproducible in the sample project in this repo.  See this patch: https://github.com/dimaMachina/graphql-eslint/issues/2260#issuecomment-2070092469

**Comment by user1736** [comment]
So for my case specifically I used the following patch:

```
diff --git a/node_modules/@graphql-eslint/eslint-plugin/cjs/graphql-config.js b/node_modules/@graphql-eslint/eslint-plugin/cjs/graphql-config.js
index 5a18a0a..00fe919
--- a/node_modules/@graphql-eslint/eslint-plugin/cjs/graphql-config.js
+++ b/node_modules/@graphql-eslint/eslint-plugin/cjs/graphql-config.js
@@ -37,7 +37,7 @@ var import_code_file_loader = require("@graphql-tools/code-file-loader");
 var import_debug = __toESM(require("debug"));
 var import_graphql_config = require("graphql-config");
 const debug = (0, import_debug.default)("graphql-eslint:graphql-config");
-let graphQLConfig;
+const configCache = new Map();
 function loadOnDiskGraphQLConfig(filePath) {
   return (0, import_graphql_config.loadConfigSync)({
     // load config relative to the file being linted
@@ -47,9 +47,12 @@ function loadOnDiskGraphQLConfig(filePath) {
     extensions: [codeFileLoaderExtension]
   });
 }
+
 function loadGraphQLConfig(options) {
-  if (graphQLConfig) {
-    return graphQLConfig;
+  // FIXME: Requires revisiting if we stop using parserOptions in eslintrc.js to feed the plugin schema and document files
+  const cacheKey = `${options.schema}#${options.documents?.length}`;
+  if (configCache.has(cacheKey)) {
+    return configCache.get(cacheKey);
   }
   const onDiskConfig = !options.skipGraphQLConfig && loadOnDiskGraphQLConfig(options.filePath);
   debug("options.skipGraphQLConfig: %o", options.skipGraphQLConfig);
@@ -64,13 +67,14 @@ function loadGraphQLConfig(options) {
     include: options.include,
     exclude: options.exclude
   };
-  graphQLConfig = onDiskConfig || new import_graphql_config.GraphQLConfig(
+  const graphQLConfig = onDiskConfig || new import_graphql_config.GraphQLConfig(
     {
       config: configOptions,
       filepath: "virtual-config"
     },
     [codeFileLoaderExtension]
   );
+  configCache.set(cacheKey, graphQLConfig);
   return graphQLConfig;
 }
 const codeFileLoaderExtension = (api) => {
```

For my configuration having schema name + number of documents is enough to distinguish between config sections, but it's not going to fly as general solution. From my testing disabling cache all together is also not an option since everything slows down to a crawl.

**Comment by superpowered** [comment]
Hello! 

I've also just run into this issue on an internal company project.

Updating: 
```
  if (process.env.NODE_ENV !== 'test' && graphQLConfig) {
    return graphQLConfig;
  }
```
to:
```
  if (process.env.NODE_ENV !== 'test' && graphQLConfig && graphQLConfig['_rawConfig'].schema === options.schema) {
    return graphQLConfig;
  }
 ```
 
Seems to be enough to fix it in our use-case at least. But I could see a test for both `.schema` and `.documents` matching to be a bit more accurate.

**Comment by dimaMachina** [comment]
@benasher44 can confirm that `no-unreachable-types` and `no-unused-fields` were broken for multi schema projects. Fixed in `@graphql-eslint/eslint-plugin@4.0.0-alpha.2` https://github.com/dimaMachina/graphql-eslint/releases/tag/release-1722466173246

**Comment by unrevised6419** [comment]
In my case it seems I still encounter this issue. When linting each project separately everything works fine. When I have the configs of both of them enabled, I get issues.

**eslint.config.mjs**

```ts
import process from "node:process";
import { defineConfig } from "eslint/config";
import graphqlPlugin from "@graphql-eslint/eslint-plugin";

const projects = [
	{
		files: ["**/*.api.gql"],
		documents: "./src/**/*.api.gql",
		schema: process.env.GRAPHQL_URL_API,
	},
	{
		files: ["**/*.cms.gql"],
		documents: "./src/**/*.cms.gql",
		schema: process.env.GRAPHQL_URL_CMS,
	},
];

export default defineConfig(
	projects.map((project) => ({
		files: project.files,
		plugins: { "@graphql-eslint": graphqlPlugin },
		languageOptions: {
			parser: graphqlPlugin.parser,
			parserOptions: {
				graphQLConfig: {
					schema: project.schema,
					documents: project.documents,
				},
			},
		},
		rules: {
			...graphqlPlugin.configs["flat/operations-all"].rules,
			"@graphql-eslint/match-document-filename": "off",
			"@graphql-eslint/alphabetize": "off",
			"@graphql-eslint/no-one-place-fragments": "off",
			"@graphql-eslint/lone-executable-definition": "off",
			"@graphql-eslint/require-import-fragment": "off",
		},
	})),
);
```

**Logs**

```
warning [graphql-eslint] Unable to locate fragment named "Page", please make sure it's loaded using "parserOptions.operations"
TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Rule "selection-set-depth" check failed due to a missing siblings operations. See https://the-guild.dev/graphql/eslint/docs/usage#providing-operations for more info TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Unable to locate fragment named "Page", please make sure it's loaded using "parserOptions.operations"
TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Rule "selection-set-depth" check failed due to a missing siblings operations. See https://the-guild.dev/graphql/eslint/docs/usage#providing-operations for more info TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Unable to locate fragment named "Page", please make sure it's loaded using "parserOptions.operations"
TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Rule "selection-set-depth" check failed due to a missing siblings operations. See https://the-guild.dev/graphql/eslint/docs/usage#providing-operations for more info TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Unable to locate fragment named "Error", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "SharedRichText", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "SharedQuote", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "SharedMedia", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "SharedSlider", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "SharedFaq", please make sure it's loaded using "parserOptions.operations"
warning [graphql-eslint] Unable to locate fragment named "Page", please make sure it's loaded using "parserOptions.operations"
TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Rule "selection-set-depth" check failed due to a missing siblings operations. See https://the-guild.dev/graphql/eslint/docs/usage#providing-operations for more info TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Unable to locate fragment named "Page", please make sure it's loaded using "parserOptions.operations"
TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)
warning [graphql-eslint] Rule "selection-set-depth" check failed due to a missing siblings operations. See https://the-guild.dev/graphql/eslint/docs/usage#providing-operations for more info TypeError: Cannot read properties of undefined (reading 'kind')
    the-project/node_modules/graphql-depth-limit/index.js:60:16)
    the-project/node_modules/graphql-depth-limit/index.js:72:14)
    the-project/node_modules/graphql-depth-limit/index.js:69:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:68:59)
    the-project/node_modules/graphql-depth-limit/index.js:77:9
    at Array.map (<anonymous>)
    the-project/node_modules/graphql-depth-limit/index.js:76:55)
    the-project/node_modules/graphql-depth-limit/index.js:22:27
    the-project/node_modules/@graphql-eslint/eslint-plugin/esm/rules/selection-set-depth/index.js:96:11)

the-project/src/app/[lang]/(web)/docs/query.cms.gql
  1:25  error  Unknown type "I18NLocaleCode"               @graphql-eslint/known-type-names
  2:2   error  Cannot query field "pages" on type "Query"  @graphql-eslint/fields-on-correct-type
  7:6   error  Unknown fragment "Page"                     @graphql-eslint/known-fragment-names

the-project/src/app/[lang]/(web)/faq/query.cms.gql
  1:24  error  Unknown type "I18NLocaleCode"               @graphql-eslint/known-type-names
  2:2   error  Cannot query field "pages" on type "Query"  @graphql-eslint/fields-on-correct-type
  7:6   error  Unknown fragment "Page"                     @graphql-eslint/known-fragment-names

the-project/src/app/[lang]/(web)/fees/query.cms.gql
  1:25  error  Unknown type "I18NLocaleCode"               @graphql-eslint/known-type-names
  2:2   error  Cannot query field "pages" on type "Query"  @graphql-eslint/fields-on-correct-type
  7:6   error  Unknown fragment "Page"                     @graphql-eslint/known-fragment-names

the-project/src/app/[lang]/(web)/privacy/query.cms.gql
  1:28  error  Unknown type "I18NLocaleCode"               @graphql-eslint/known-type-names
  2:2   error  Cannot query field "pages" on type "Query"  @graphql-eslint/fields-on-correct-type
  7:6   error  Unknown fragment "Page"                     @graphql-eslint/known-fragment-names

the-project/src/app/[lang]/(web)/terms/query.cms.gql
  1:26  error  Unknown type "I18NLocaleCode"               @graphql-eslint/known-type-names
  2:2   error  Cannot query field "pages" on type "Query"  @graphql-eslint/fields-on-correct-type
  7:6   error  Unknown fragment "Page"                     @graphql-eslint/known-fragment-names

the-project/src/app/[lang]/page.fragment.cms.gql
   1:1   error  Fragment "Page" is never used              @graphql-eslint/no-unused-fragments
   1:18  error  Unknown type "Page". Did you mean "Date"?  @graphql-eslint/known-type-names
   7:10  error  Unknown type "Error"                       @graphql-eslint/known-type-names
  10:10  error  Unknown type "ComponentSharedRichText"     @graphql-eslint/known-type-names
  13:10  error  Unknown type "ComponentSharedQuote"        @graphql-eslint/known-type-names
  16:10  error  Unknown type "ComponentSharedMedia"        @graphql-eslint/known-type-names
  19:10  error  Unknown type "ComponentSharedSlider"       @graphql-eslint/known-type-names
  22:10  error  Unknown type "ComponentSharedFaq"          @graphql-eslint/known-type-names
  28:1   error  Fragment "Error" is never used             @graphql-eslint/no-unused-fragments
  28:19  error  Unknown type "Error"                       @graphql-eslint/known-type-names
  33:1   error  Fragment "SharedRichText" is never used    @graphql-eslint/no-unused-fragments
  33:28  error  Unknown type "ComponentSharedRichText"     @graphql-eslint/known-type-names
  38:1   error  Fragment "SharedQuote" is never used       @graphql-eslint/no-unused-fragments
  38:25  error  Unknown type "ComponentSharedQuote"        @graphql-eslint/known-type-names
  43:1   error  Fragment "SharedMedia" is never used       @graphql-eslint/no-unused-fragments
  43:25  error  Unknown type "ComponentSharedMedia"        @graphql-eslint/known-type-names
  48:1   error  Fragment "SharedSlider" is never used      @graphql-eslint/no-unused-fragments
  48:26  error  Unknown type "ComponentSharedSlider"       @graphql-eslint/known-type-names
  53:1   error  Fragment "SharedFaq" is never used         @graphql-eslint/no-unused-fragments
  53:23  error  Unknown type "ComponentSharedFaq"          @graphql-eslint/known-type-names

✖ 35 problems (35 errors, 0 warnings)
```
